torch/_functorch/partitioners.py

```diff
@@ -2598,6 +2598,88 @@ def has_same_nodes(joint_graph):
     return saved_values
 
 
+def thread_graphsafe_rng_from_hops(module, is_backward):
+    """
+    Graph-safe RNG lets torch.compile use CUDA Graphs for graphs with RNG ops.
+    For graphs without HOPs, the partitioner adds placeholder nodes
+    fwd_rng_state_* and bw_rng_state_* to the forward and backward graphs. At
+    runtime, the AOTDispatcher retrieves these RNG states and passes them to the
+    compiled graphs.
+
+    This works well for no-HOP graphs. With HOPs, the partitioner runs
+    recursively: it first partitions the HOP (producing forward/backward HOP
+    subgraphs) and then stitches them back into the outer joint graph. For HOPs
+    that contain RNG ops, the outer joint graph now includes HOP subgraph
+    modules with extra RNG placeholders. We must thread these placeholders
+    through the outer module partitioned forward and backward graphs—this
+    function does exactly that. It collects the RNG placeholder nodes from the
+    HOPs and creates corresponding placeholders in the outer forward and
+    backward graphs.
+
+    There is a catch: for a short period, the joint graph is in a “bad” state.
+    The HOP subgraphs expect additional inputs (because of the new
+    placeholders), but the outer graph call sites don't yet provide them. We
+    can't fix this in the joint graph because the joint graph's input signature
+    is fixed (primals, tangents). As a compromise, we keep the joint graph in
+    somewhat of a bad state for some time and, once the outer forward and
+    backward graphs are partitioned, insert the corresponding RNG placeholders
+    and wire up the calls.
```

Comment:
Would it be clearer to have a temporary node that represents the yet to be added placeholder node ? Then both the non-hop and hop could have a pass to lift them to placeholders ?

Reply:
I thought about that a little more, but I could not figure out how to add that temporary node. It can not be a placholder on the joint graph because the signature is `primals, tangents` where both of them are lists. And other parts of the stack (mostly partitioner) assumes this signature. If we make it (primals, tangents, fwd_rng_state, bw_rng_state), we will have to make changes at many many places. At that point, it might end up being more hacky.